### ISSUE TITLE
Yolov5 performs not well in my object detection,many object not been detected,just been considered as background？

### ISSUE THREAD

**Issue by wyh196646** [ultralytics/yolov5#7981]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Yolov5 performs not well in my object detection,many object not been detected,just been considered as background？
I just use yolov5 for chest X-ray abnormality detection,the kaggle link is [https://www.kaggle.com/competitions/vinbigdata-chest-xray-abnormalities-detection](url) ,I just get the confusion matrix like this
![confusion_matrix](https://user-images.githubusercontent.com/42706208/170301835-c190d6b3-1076-40b0-9517-4ef6d1649bff.png)

this show that many abnormality was not detected by the model，I konw the dataset is not imbalanced,and some object is really small，all in all,the problem in the dataset is really severe.
here is some result file of my train pipeline

![F1_curve](https://user-images.githubusercontent.com/42706208/170302585-4c4d1813-2661-4875-b5e5-eb6d90d2b121.png)
![labels](https://user-images.githubusercontent.com/42706208/170302603-92ed4266-ec12-458d-abd5-e2902319ded3.jpg)
![labels_correlogram](https://user-images.githubusercontent.com/42706208/170302607-dd1851a9-8a51-4876-a45e-16da47d69fef.jpg)
![P_curve](https://user-images.githubusercontent.com/42706208/170302615-2c3b824a-336e-4d4f-a316-40b926a35d06.png)
![PR_curve](https://user-images.githubusercontent.com/42706208/170302620-ac4142bc-94ec-435f-a4e2-90c8686bbba6.png)
![R_curve](https://user-images.githubusercontent.com/42706208/170302622-a4ac1175-7936-43bc-9683-0bcaf4a89bea.png)
![results](https://user-images.githubusercontent.com/42706208/170302629-02726cd4-8302-460c-9abd-bfb9c58b8edd.png)
![train_batch0](https://user-images.githubusercontent.com/42706208/170302633-5bad0abe-b3ab-4c18-bcc7-cc85d65b3f1a.jpg)
![train_batch1](https://user-images.githubusercontent.com/42706208/170302643-bd5ff385-aa6c-4275-8a03-f59749373aed.jpg)
![train_batch2](https://user-images.githubusercontent.com/42706208/170302652-59563c23-8212-4e2d-8cbd-49606c617e45.jpg)
![val_batch0_labels](https://user-images.githubusercontent.com/42706208/170302664-94269398-6710-4a8d-a376-19b9074892bc.jpg)
![val_batch0_pred](https://user-images.githubusercontent.com/42706208/170302675-7445b484-8074-46f5-91e4-bec5ebf779ed.jpg)
![val_batch1_labels](https://user-images.githubusercontent.com/42706208/170302697-5007bdeb-b694-4832-9fe1-f15e58003a6a.jpg)
![val_batch1_pred](https://user-images.githubusercontent.com/42706208/170302706-66ae658d-8e31-4c10-8047-6e117aeb2133.jpg)
![val_batch2_labels](https://user-images.githubusercontent.com/42706208/170302717-53b8c6cd-3e59-4765-8f13-72e866de2ee0.jpg)
![val_batch2_pred](https://user-images.githubusercontent.com/42706208/170302725-1ecc7c23-12f3-486e-92ac-a77af4612b52.jpg)

I sincerely hope to get some useful suggestions, no matter about model improvement or data processing, hope to get everyone's suggestions and help!

**Comment by glenn-jocher** [comment]
@wyh196646 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by SaadTazroute** [comment]
I suggest that you take a look of the focal loss. Try different values of gamma between [1,5], it could help conerning your imbalanced classes

**Comment by glenn-jocher** [comment]
@SaadTazroute thanks for the suggestion and for bringing up the focal loss! As you pointed out, focal loss, with its modifiable gamma parameter, can be quite effective for addressing imbalanced classes. We’ve implemented focal loss in YOLOv5 and default to `--fl-gamma 0.0` for balanced loss, but you can tune it as needed, typically higher for harder datasets.

Happy training!
